endpoints/v2/manifest.py

```diff
@@ -47,129 +57,189 @@
+                # or if it's an org that is not set up as a cache.
+                # either way we want the regular code to do its thing, so we
+                # give over control.
+                pass
+
+        return self.get_manifest_by_ref(namespace_name, repo_name, manifest_ref)
+
+    def get_manifest_by_ref(namespace_name, repo_name, manifest_ref):
+        raise NotImplementedError()
+
+    def _proxy_cache(self, config, namespace_name, repo_name, manifest_ref):
+        # users can configure a cache org to proxy a whole registry or a single namespace.
+        # when proxying an entire registry, the upstream namespace is not set, and the
+        # given repo_name will be an exact match of the upstream namespace/repository combo.
+        repo = repo_name
+        target_ns = config.upstream_registry_namespace
+        if target_ns != "" and target_ns is not None:
+            # when proxying a single namespace, the repo_name will not contain the upstream namespace,
+            # so we get that from the proxy config and put them together.
+            repo = f"{target_ns}/{repo_name}"
+
+        auth = None
+        username = config.upstream_registry_username.decrypt()
+        password = config.upstream_registry_password.decrypt()
+        if username is not None and password is not None:
+            auth = (username, password)
+
+        proxy = Proxy(config.upstream_registry, repo, auth)
+        media_type = request.headers.get("Accept", None)
+        resp = proxy.get_manifest(manifest_ref, media_type)
+        return Response(**resp)
+
+
+class FetchManifestByTag(FetchManifest):
+
+    def get_manifest_by_ref(namespace_name, repo_name, manifest_ref):
+        repository_ref = registry_model.lookup_repository(namespace_name, repo_name)
+        if repository_ref is None:
+            image_pulls.labels("v2", "tag", 404).inc()
+            raise NameUnknown()
+
+        tag = registry_model.get_repo_tag(repository_ref, manifest_ref)
+        if tag is None:
+            if registry_model.has_expired_tag(repository_ref, manifest_ref):
+                logger.debug(
+                    "Found expired tag %s for repository %s/%s", manifest_ref, namespace_name, repo_name
+                )
+                msg = (
+                    "Tag %s was deleted or has expired. To pull, revive via time machine" % manifest_ref
+                )
+                image_pulls.labels("v2", "tag", 404).inc()
+                raise TagExpired(msg)
 
-    tag = registry_model.get_repo_tag(repository_ref, manifest_ref)
-    if tag is None:
-        if registry_model.has_expired_tag(repository_ref, manifest_ref):
-            logger.debug(
-                "Found expired tag %s for repository %s/%s", manifest_ref, namespace_name, repo_name
-            )
-            msg = (
-                "Tag %s was deleted or has expired. To pull, revive via time machine" % manifest_ref
-            )
```

Comment:
Won't removing this be a bad UX design for deleted tags that are in normal, i.e. not proxied repositories? Time machine is still a thing, we should give the user info that a tag is in time machine if the feature is enabled.

Reply:
Not sure I'm following. This was just moved around: https://github.com/quay/quay/pull/1053/files#diff-2487e826aa015b45ecc8a34cbf59ea7bccf235b8d51bf732fd574cf94d11eeedR132

Reply:
I missed the new location, carry on :-)

Reply:
no worries, diffs can be confusing! I tried the unified view but it was worse, might be something worth to play with tho.